src/mongocxx/events/command_failed_event.hpp

```diff
@@ -75,6 +79,15 @@ class MONGOCXX_API command_failed_event {
     ///
     std::int64_t operation_id() const;
 
+    ///
+    /// Returns the service id.
+    ///
+    /// @note service_id() only returns a value if load balancing is enabled.
+    ///
+    /// @return No value, or the service id if load balancing is enabled.
+    ///
+    bsoncxx::stdx::optional<bson_oid_t> service_id() const;
```

Comment:
This cannot return a libbson type. To return a BSON ObjectID, you can use the `bsoncxx:oid`. For some prior art, see `mongocxx::events::server_changed_event::topology_id()`, which also returns a BSON ObjectID in public API by converting from a libbson `bson_oid_t`.

Aside: https://bsonspec.org/spec.html is the canonical reference on BSON types.

Reply:
Ok, thank you!! I'll rework this.